serving/src/main/java/feast/serving/service/OnlineServingService.java

```diff
@@ -64,110 +62,211 @@ public GetFeastServingInfoResponse getFeastServingInfo(
-        FeatureSetRequest featureSetRequest = featureSetRequests.get(fsIdx);
-
-        String project = featureSetRequest.getSpec().getProject();
-
-        // In order to return values containing the same feature references provided by the user,
-        // we reuse the feature references in the request as the keys in the featureValuesMap
-        Map<String, FeatureReference> refsByName = featureSetRequest.getFeatureRefsByName();
-
-        // Each feature row returned (per feature set request) corresponds to a given entity row.
-        // For each feature row, update the featureValuesMap.
-        for (var entityRowIdx = 0; entityRowIdx < entityRows.size(); entityRowIdx++) {
-          FeatureRow featureRow = featureRowsForFs.get(entityRowIdx);
-          EntityRow entityRow = entityRows.get(entityRowIdx);
-
-          // If the row is stale, put an empty value into the featureValuesMap.
-          if (isStale(featureSetRequest, entityRow, featureRow)) {
-            featureSetRequest
-                .getFeatureReferences()
-                .forEach(
-                    ref -> {
-                      populateStaleKeyCountMetrics(project, ref);
-                      featureValuesMap
-                          .get(entityRow)
-                          .put(RefUtil.generateFeatureStringRef(ref), Value.newBuilder().build());
-                    });
-
-          } else {
-            populateRequestCountMetrics(featureSetRequest);
-
-            // Else populate the featureValueMap at this entityRow with the values in the feature
-            // row.
-            featureRow.getFieldsList().stream()
-                .filter(field -> refsByName.containsKey(field.getName()))
-                .forEach(
-                    field -> {
-                      FeatureReference ref = refsByName.get(field.getName());
-                      String id = RefUtil.generateFeatureStringRef(ref);
-                      featureValuesMap.get(entityRow).put(id, field.getValue());
-                    });
-          }
+      List<FeatureSetRequest> featureSetRequests =
+          specService.getFeatureSets(request.getFeaturesList());
+      for (FeatureSetRequest featureSetRequest : featureSetRequests) {
+        // Pull feature rows for given entity rows from the feature/featureset specified in feature
+        // set request.
+        // from the configured online
+        List<FeatureRow> featureRows = retriever.getOnlineFeatures(entityRows, featureSetRequest);
+        // Check that feature row returned corresponds to a given entity row.
+        if (featureRows.size() != entityRows.size()) {
+          throw Status.INTERNAL
+              .withDescription(
+                  "The no. of FeatureRow obtained from OnlineRetriever"
+                      + "does not match no. of entityRow passed.")
+              .asRuntimeException();
         }
+
+        for (var i = 0; i < entityRows.size(); i++) {
```

Comment:
Can we use a more modern means of iteration for readability?

Reply:
Changed.